File: src/so/aporia/module/impl/misc/PacketDebug.java
/*
 * Copyright (c) 2025-2026 Aporia.cc Project
 * Distributed under the Aporia.cc Software License Agreement v1.0
 * See LICENSE and COPYRIGHT files in the project root for full text.
 */

package so.aporia.module.impl.misc;

import net.minecraft.network.protocol.Packet;
import so.aporia.module.Category;
import so.aporia.module.Module;
import so.aporia.module.settings.BooleanSetting;
import so.aporia.utils.events.EventBus;
import so.aporia.utils.events.EventHandler;
import so.aporia.utils.events.impl.PacketEvent;
import so.aporia.utils.files.FilesManager;
import so.aporia.utils.user.logger.Logger;

import java.nio.file.Path;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;

/**
 * PacketDebug — логирует ВСЕ пакеты в .log файл (append).
 * При каждом enable создаёт новый parse_<timestamp>.log
 */
public final class PacketDebug extends Module {

    public final BooleanSetting logInbound = new BooleanSetting(
        "Log Inbound", "Входящие пакеты", true
    );
    public final BooleanSetting logOutbound = new BooleanSetting(
        "Log Outbound", "Исходящие пакеты", true
    );

    private static final DateTimeFormatter TIME_FMT =
        DateTimeFormatter.ofPattern("HH:mm:ss.SSS");
    private static final DateTimeFormatter FILE_FMT =
        DateTimeFormatter.ofPattern("yyyy-MM-dd_HH-mm-ss");

    private Path sessionFile;
    private long packetCount = 0;

    public PacketDebug() {
        super("PacketDebug", Category.MISC);
    }

    @Override
    protected void onEnable() {
        packetCount = 0;
        String ts = LocalDateTime.now().format(FILE_FMT);
        sessionFile = FilesManager.ROOT.resolve("packets/parse_" + ts + ".log");
        try {
            String header = "=== PacketDebug Session: " + ts + " ===\n\n";
            FilesManager.append(sessionFile, header);
            Logger.success("[PacketDebug] ENABLED → " + sessionFile.getFileName());
        } catch (Exception e) {
            Logger.error("[PacketDebug] init failed: " + e.getMessage());
            sessionFile = null;
            return;
        }
        EventBus.INSTANCE.register(this);
    }

    @Override
    protected void onDisable() {
        EventBus.INSTANCE.unregister(this);
        try {
            String footer = "\n=== Session ended (" + packetCount + " packets) ===\n";
            FilesManager.append(sessionFile, footer);
            Logger.success("[PacketDebug] DISABLED — logged " + packetCount + " packets");
        } catch (Exception e) {
            Logger.error("[PacketDebug] disable failed: " + e.getMessage());
        }
        sessionFile = null;
    }

    @EventHandler
    public void onPacket(PacketEvent event) {
        if (sessionFile == null) return;
        if (event.direction() == PacketEvent.Direction.INBOUND  && !logInbound.isEnabled())  return;
        if (event.direction() == PacketEvent.Direction.OUTBOUND && !logOutbound.isEnabled()) return;
        Packet<?> pkt = event.packet();
        String simple = pkt.getClass().getSimpleName();
        String dir = event.direction() == PacketEvent.Direction.INBOUND ? "IN " : "OUT";
        String time = LocalDateTime.now().format(TIME_FMT);
        String entry = '[' + time + "] [" + dir + "] " + simple + '\n'
            + "  class: " + pkt.getClass().getName() + '\n'
            + "  data : " + pkt + '\n'
            + "──────────────────────────────────────────\n";
        try {
            FilesManager.append(sessionFile, entry);
            packetCount++;
        } catch (Exception e) {
            Logger.error("[PacketDebug] write failed: " + e.getMessage());
        }
    }
}

File: src/so/aporia/module/impl/misc/PacketDebug.kt
package so.aporia.module.impl.misc
import com.chaos.annotation.Obfuscate
import net.minecraft.network.protocol.Packet
import net.minecraft.network.protocol.game.ClientboundCommandsPacket
import net.minecraft.network.protocol.game.ClientboundSystemChatPacket
import net.minecraft.network.protocol.common.ClientboundCustomPayloadPacket
import net.minecraft.network.protocol.common.custom.BrandPayload
import net.minecraft.network.protocol.game.ServerboundCommandSuggestionPacket
import so.aporia.module.Category
import so.aporia.module.Module
import so.aporia.module.settings.BooleanSetting
import so.aporia.utils.events.EventHandler
import so.aporia.utils.events.impl.PacketEvent
import so.aporia.utils.files.FilesManager
import so.aporia.utils.imports.*
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter
import com.chaos.annotation.ChaosNative
@Obfuscate
@ChaosNative
class PacketDebug : Module("PacketDebug", Category.MISC) {

    val logInbound = BooleanSetting("Log Inbound", "Входящие пакеты", true)
    val logOutbound = BooleanSetting("Log Outbound", "Исходящие пакеты", true)

    private var sessionFile: java.nio.file.Path? = null
    private var packetCount = 0L

    override fun onEnable() {
        packetCount = 0
        val ts = LocalDateTime.now().format(FILE_FMT)
        sessionFile = FilesManager.ROOT.resolve("packets/parse_$ts.log")
        try {
            FilesManager.append(sessionFile!!, "=== PacketDebug Session: $ts ===\n\n")
            logger.success("[PacketDebug] ENABLED → ${sessionFile!!.fileName}")
        } catch (e: Exception) {
            logger.error("[PacketDebug] init failed: ${e.message}")
            sessionFile = null
            return
        }
        bus.register(this)
    }

    override fun onDisable() {
        bus.unregister(this)
        try {
            val sf = sessionFile ?: return
            FilesManager.append(sf, "\n=== Session ended ($packetCount packets) ===\n")
            logger.success("[PacketDebug] DISABLED — logged $packetCount packets")
        } catch (e: Exception) {
            logger.error("[PacketDebug] disable failed: ${e.message}")
        }
        sessionFile = null
    }

    private fun dumpCommands(pkt: ClientboundCommandsPacket): String {
        try {
            val field = ClientboundCommandsPacket::class.java.getDeclaredField("entries")
            field.isAccessible = true
            @Suppress("UNCHECKED_CAST")
            val entries = field.get(pkt) as List<*>
            val names = mutableListOf<String>()

            for (entry in entries) {
                val e = entry ?: continue
                val stubField = e::class.java.getDeclaredField("stub")
                stubField.isAccessible = true
                val stub = stubField.get(e)
                if (stub != null) {
                    val idField = stub::class.java.getDeclaredField("id")
                    idField.isAccessible = true
                    val id = idField.get(stub) as? String
                    if (id != null) names.add(id)
                }
            }

            val sorted = names.sorted().distinct()
            return "Commands[${sorted.size}]: ${sorted.joinToString(", ")}"
        } catch (e: Exception) {
            return "CommandsPacket (parse error: ${e.message})"
        }
    }

    private fun packetData(pkt: Packet<*>): String {
        return when (pkt) {
            is ClientboundCommandsPacket -> dumpCommands(pkt)
            is ServerboundCommandSuggestionPacket -> {
                "cmd=\"${pkt.command}\""
            }
            is ClientboundSystemChatPacket -> {
                "msg=\"${pkt.content.getString()}\""
            }
            is ClientboundCustomPayloadPacket -> {
                val pl = pkt.payload()
                val chan = pl.type().id()
                if (pl is BrandPayload) "BrandPayload: \"${pl.brand}\"" else "CustomPayload: $chan"
            }
            else -> pkt.toString()
        }
    }

    @EventHandler
    fun onPacket(event: PacketEvent) {
        val sf = sessionFile ?: return
        if (event.direction == PacketEvent.Direction.INBOUND && !logInbound.isEnabled) return
        if (event.direction == PacketEvent.Direction.OUTBOUND && !logOutbound.isEnabled) return
        val pkt = event.packet
        val dir = if (event.direction == PacketEvent.Direction.INBOUND) "IN " else "OUT"
        val time = LocalDateTime.now().format(TIME_FMT)
        val data = packetData(pkt)
        val entry = "[$time] [$dir] ${pkt.javaClass.simpleName}\n  class: ${pkt.javaClass.name}\n  data : $data\n──────────────────────────────────────────\n"
        try {
            FilesManager.append(sf, entry)
            packetCount++
        } catch (e: Exception) {
            logger.error("[PacketDebug] write failed: ${e.message}")
        }
    }

    companion object {
        private val TIME_FMT = DateTimeFormatter.ofPattern("HH:mm:ss.SSS")
        private val FILE_FMT = DateTimeFormatter.ofPattern("yyyy-MM-dd_HH-mm-ss")
    }

    override val settings = listOf(logInbound, logOutbound)
}